Sources/Flows/OAuth2DeviceGrant.swift

```diff
@@ -0,0 +1,194 @@
+		try assureNoErrorInResponse(params)
+		
+		return params
+	}
+	
+	// custom
+	public func start(useNonTextualTransmission: Bool = false, completion: @escaping (DeviceAuthorization?, Error?) -> Void) {
+		authorizeDevice { result, error in
+			guard let result else {
+				if let error {
+					self.logger?.warn("OAuth2", msg: "Unable to get device code: \(error)")
+				}
+				completion(nil, error)
+				return
+			}
+			
+			guard let deviceCode = result["device_code"] as? String,let userCode = result["user_code"] as? String,
+				  let verificationUri = result["verification_uri"] as? String,
+				  let verificationUrl = URL(string: verificationUri),
+				  let expiresIn = result["expires_in"] as? Int else {
+				let error = OAuth2Error.generic("The response doesn't conttain all required fields.")
+				self.logger?.warn("OAuth2", msg: String(describing: error))
+				completion(nil, error)
+				return
+			}
+			
+			var verificationUrlComplete: URL?
+			if let verificationUriComplete = result["verification_uri_complete"] as? String {
+				verificationUrlComplete = URL(string: verificationUriComplete)
+			}
+			
+			if useNonTextualTransmission, let url = verificationUrlComplete {
+				do {
+					try self.authorizer.openAuthorizeURLInBrowser(url)
+				} catch let error {
+					completion(nil, error)
+				}
+			}
+			
+			let pollingInterval = result["interval"] as? TimeInterval ?? 5 // TODO: use Int instead?
+			self.getDeviceAccessToken(deviceCode: deviceCode, interval: pollingInterval) { params, error in
+				if let params {
+					self.didAuthorize(withParameters: params)
+				}
+				else if let error {
+					self.didFail(with: error.asOAuth2Error)
+				}
+			}
+			
+			let deviceAuthorization = DeviceAuthorization(userCode: userCode, verificationUrl: verificationUrl, verificationUrlComplete: verificationUrlComplete, expiresIn: expiresIn)
+			completion(deviceAuthorization, nil)
+		}
+	}
+	
+	private func authorizeDevice(completion: @escaping (OAuth2JSON?, Error?) -> Void) {
+		do {
+			let post = try deviceAuthorizationRequest().asURLRequest(for: self)
+			logger?.debug("OAuth2", msg: "Obtaining device code from \(post.url!)")
+			
+			perform(request: post) { response in
+				do {
+					let data = try response.responseData()
+					let params = try self.parseDeviceAuthorizationResponse(data: data)
+					completion(params, nil)
+				}
+				catch let error {
+					completion(nil, error.asOAuth2Error)
+				}
+			}
+		}
+		catch let error {
+			completion(nil, error.asOAuth2Error)
+		}
+	}
+	
+	private func getDeviceAccessToken(deviceCode: String, interval: TimeInterval, completion: @escaping (OAuth2JSON?, Error?) -> Void) {
+		do {
+			let post = try deviceAccessTokenRequest(with: deviceCode).asURLRequest(for: self)
+			logger?.debug("OAuth2", msg: "Obtaining access token for device with code \(deviceCode) from \(post.url!)")
+			
+			// TODO: check if not expired
```

Comment:
Should this be done or can this be done later?

Reply:
@DominikPalo

Reply:
@ossus-lib  Thank you, I fixed it (I just forgot to cherry-pick a local commit where I cleaned development comments, it should be ok now)